### PR TITLE
Redfish: Expanded SimpleUpdate command to allow for users to monitor the progress of an update and perform follow-up operations

### LINKED ISSUE #3910: redfish_config - set HttpPushUriOptions / atttribute ApplyTime to OnReset
### Summary

Hi all,
I may misunderstood how redfish_info module/module_utils behaves, but I can't simply set the ApplyTime attribute.

Set HttpPushUriOptions ApplyTime=OnReset 
Is it possible ? What do I miss ?

Best regards,
Francine 

### Issue Type

Feature Idea

### Component Name

redfish_config remote_management

### Additional Information

```yaml (paste below)
  community.general.redfish_config:
    category: UpdateService
    command: SetHttpPushUriOptions
    httppushurioptions:
      HttpPushUriOptions:
        HttpPushUriApplyTime
          ApplyTime: OnReset
```

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

### LINKED ISSUE #4276: Add capabilities to the Redfish SimpleUpdate redfish_command
### Summary

The Redfish Firmware Updater process described [here (See Chapter 6)](https://www.dmtf.org/sites/default/files/standards/documents/DSP2062_1.0.0.pdf) allows for asynchronous polling of the update job by querying the returned "TaskMonitor". Currently, the Redfish Command handled by:

https://github.com/ansible-collections/community.general/blob/8508e3fa6f433c74bedd8ccb1319d295485e6a7f/plugins/module_utils/redfish_utils.py#L1339-L1398

only returns True/False on the job submission, but no other details about the job. This adds extra complexity in searching for the Task after the fact and manually polling.

It would be great if the RF utils module could return Task Monitors where appropriate for long running jobs and It would be even more awesome if the module could abstract the async job process away from the user:

- Accept a "Wait for completion" parameter that would instruct the module to poll the Task and return status.
- Accept the `@Redfish.OperationApplyTime` parameter for more control over the update process timing.

See also [Redfish Task Presentation](https://www.dmtf.org/sites/default/files/Redfish%20School%20-%20Tasks_1.pdf)

### Issue Type

Feature Idea

### Component Name

Redfish Utils

### Additional Information

When using the URI module a SimpleUpdate returns:

```
{
...
  "json": {
    "@odata.context": "/redfish/v1/$metadata#Task.Task",
    "@odata.id": "/redfish/v1/TaskService/Tasks/2",
    "@odata.type": "#Task.v1_3_0.Task",
    "Description": "Update Task",
    "Id": "2",
    "Messages": [
      {}
    ],
    "Name": "Task 2",
    "Payload": {
      "HttpOperation": "POST",
      "JsonBody": "{\"ImageURI\":\"http://updatefile\"}",
      "TargetUri": "/redfish/v1/UpdateService/Actions/UpdateService.SimpleUpdate"
    },
    "StartTime": "2022-02-07T17:12:41Z",
    "TaskMonitor": "/redfish/v1/TaskService/TaskMonitors/2",
    "TaskState": "New"
  },
  "location": "https://1.2.3.4/redfish/v1/TaskService/TaskMonitors/2",
  "msg": "OK (unknown bytes)",
...
}
```

Where the `TaskMonitor` and `TaskState` are returned for further use.

```
The Redfish Command SimpleUpdate only returns:
TASK [Simple update] ************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************
changed: [localhost] => {"attempts": 1, "changed": true, "msg": "Action was successful", "session": {}}

TASK [debug] ********************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "bios_upgrade": {
        "attempts": 1,
        "changed": true,
        "failed": false,
        "msg": "Action was successful",
        "session": {}
    }
}
```

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

### PR DESCRIPTION
**Pull request by mraineri**
##### SUMMARY

The Redfish Forum published a firmware update white paper last year that walked through a standard end-to-end workflow for performing firmware updates. This change expands the Redfish module in a few areas:

* Allows an "operation apply time" parameter to be specified as part of the SimpleUpdate command
* Adds response data to the SimpleUpdate command
    * The response data contains a "handle" used by a user to monitor the progress of the update
    * It also contains other status information about the update itself
* Adds a GetUpdateStatus command to check the progress of an update request
    * The "handle" obtained from the SimpleUpdate request or the previous GetUpdateStatus command is passed in
    * The response contains the same status structure returned from SimpleUpdate; it's possible the handle could move elsewhere
* Adds a PerformRequestedOperations command to perform anything necessary to continue the update flow
    * Updates can pause in a "UserIntervention" state, meaning the user needs to do some operation for the update to continue
    * These operations can perform resets of devices, so a user is needed here to let the API know it's safe to continue updating

Fixes #3910 
Fixes #4276

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

redfish_utils, redfish_command, redfish_info

##### ADDITIONAL INFORMATION

The following is a general flow of the end-to-end update process:

```
update_status = SimpleUpdate()
while( update_status['handle'] ) {
    update_status = CheckUpdateStatus( handle )
    if update_status['status'] == 'UserIntervention' {
        PerformRequestedOperations( update_status['handle'] )
    }
}
print_update_results( update_status )
```

First step, perform a simple update; the following is the playbook along with output:

```
--
- hosts: all
  gather_facts: false
  vars:
    username: <REDACTED>
    password: <REDACTED>
    baseuri: <REDACTED>
    default_uri_timeout: 5
    default_uri_retries: 5
  tasks:
  - name: Perform simple update
    community.general.redfish_command:
      category: Update
      command: SimpleUpdate
      baseuri: "{{ baseuri }}"
      username: "{{ username }}"
      password: "{{ password }}"
      update_image_uri: http://<REDACTED>/firmimgFIT.d9
    retries: "{{ default_uri_retries }}"
    register: redfish_results
  - debug:
      var: redfish_results


PLAY [all] ********************************************************************************************************************************************

TASK [Perform simple update] **************************************************************************************************************************
changed: [localhost]

TASK [debug] ******************************************************************************************************************************************
ok: [localhost] => {
    "redfish_results": {
        "ansible_facts": {
            "discovered_interpreter_python": "/usr/bin/python3"
        },
        "changed": true,
        "failed": false,
        "msg": "Action was successful",
        "redfish_facts": {
            "update_status": {
                "handle": "/redfish/v1/TaskService/Tasks/JID_687419710670",
                "messages": [],
                "resets_requested": [],
                "ret": true,
                "status": "New"
            }
        },
        "session": {}
    }
}

PLAY RECAP ********************************************************************************************************************************************
localhost                  : ok=2    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0 
```

Next is polling the update status. The following playbook was performed twice. The first time the update was pending, and the second time shows the update is complete. Notice the "handle" comes back as null in the second case, and the status shows "Completed".

```
---
- hosts: all
  gather_facts: false
  vars:
    username: <REDACTED>
    password: <REDACTED>
    baseuri: <REDACTED>
    default_uri_timeout: 5
    default_uri_retries: 5
  tasks:
  - name: Get Update Status
    community.general.redfish_info:
      category: Update
      command: GetUpdateStatus
      baseuri: "{{ baseuri }}"
      username: "{{ username }}"
      password: "{{ password }}"
      update_handle: /redfish/v1/TaskService/Tasks/JID_687419710670
    retries: "{{ default_uri_retries }}"
    register: redfish_results
  - debug:
      var: redfish_results


PLAY [all] ********************************************************************************************************************************************

TASK [Get Update Status] ******************************************************************************************************************************
ok: [localhost]

TASK [debug] ******************************************************************************************************************************************
ok: [localhost] => {
    "redfish_results": {
        "ansible_facts": {
            "discovered_interpreter_python": "/usr/bin/python3"
        },
        "changed": false,
        "failed": false,
        "redfish_facts": {
            "update_status": {
                "handle": "/redfish/v1/TaskService/Tasks/JID_687419710670",
                "messages": [
                    {
                        "Message": "Downloading the firmimgFIT.d9 update package.",
                        "MessageArgs": [],
                        "MessageArgs@odata.count": 0,
                        "MessageId": "IDRAC.2.5.RED110"
                    }
                ],
                "resets_requested": [],
                "ret": true,
                "status": "Pending"
            }
        }
    }
}

PLAY RECAP ********************************************************************************************************************************************
localhost                  : ok=2    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   


PLAY [all] ********************************************************************************************************************************************

TASK [Get Update Status] ******************************************************************************************************************************
ok: [localhost]

TASK [debug] ******************************************************************************************************************************************
ok: [localhost] => {
    "redfish_results": {
        "ansible_facts": {
            "discovered_interpreter_python": "/usr/bin/python3"
        },
        "changed": false,
        "failed": false,
        "redfish_facts": {
            "update_status": {
                "handle": null,
                "messages": [
                    {
                        "Message": "Job completed successfully.",
                        "MessageArgs": [],
                        "MessageArgs@odata.count": 0,
                        "MessageId": "IDRAC.2.5.RED001"
                    }
                ],
                "resets_requested": [],
                "ret": true,
                "status": "Completed"
            }
        }
    }
}

PLAY RECAP ********************************************************************************************************************************************
localhost                  : ok=2    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```

Finally, although not operations are needed in this case (I could not find a system that required follow-up steps at this time), the previous handle is still used to perform requested operations; the result shows nothing was done (changed = False).

```
---
- hosts: all
  gather_facts: false
  vars:
    username: <REDACTED>
    password: <REDACTED>
    baseuri: <REDACTED>
    default_uri_timeout: 5
    default_uri_retries: 5
  tasks:
  - name: Perform operations to continue the update
    community.general.redfish_command:
      category: Update
      command: PerformRequestedOperations
      baseuri: "{{ baseuri }}"
      username: "{{ username }}"
      password: "{{ password }}"
      update_handle: /redfish/v1/TaskService/Tasks/JID_687419710670
    retries: "{{ default_uri_retries }}"
    register: redfish_results
  - debug:
      var: redfish_results


PLAY [all] ********************************************************************************************************************************************

TASK [Perform operations to continue the update] ******************************************************************************************************
ok: [localhost]

TASK [debug] ******************************************************************************************************************************************
ok: [localhost] => {
    "redfish_results": {
        "ansible_facts": {
            "discovered_interpreter_python": "/usr/bin/python3"
        },
        "changed": false,
        "failed": false,
        "msg": "Action was successful",
        "redfish_facts": {},
        "session": {}
    }
}

PLAY RECAP ********************************************************************************************************************************************
localhost                  : ok=2    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```

### CODE REVIEW

File: plugins/modules/redfish_command.py [hunk 1]
```diff
@@ -859,14 +886,19 @@ def main():
...
 
     # Return data back or fail with proper message
     if result['ret'] is True:
         del result['ret']
         changed = result.get('changed', True)
         session = result.get('session', dict())
         module.exit_json(changed=changed, session=session,
-                         msg='Action was successful')
+                         msg='Action was successful',
+                         redfish_facts=redfish_facts)
```
Comment (mraineri) [inline plugins/modules/redfish_command.py:901]: Is this addition okay? Most commands will not have any output to redfish_facts, and it doesn't seem normal from a "command" perspective to have output data like this, but I couldn't think of other ways to encapsulate this information.

File: plugins/modules/redfish_command.py [hunk 2]
```diff
@@ -859,14 +886,19 @@ def main():
         for command in command_list:
             if command == "SimpleUpdate":
                 result = rf_utils.simple_update(update_opts)
+                if 'update_status' in result:
+                    redfish_facts['update_status'] = result['update_status']
+            elif command == "PerformRequestedOperations":
```
Comment (mraineri) [inline plugins/modules/redfish_command.py:891]: I'm certainly open to opinions about the name if anyone has suggestions.

File: plugins/modules/redfish_command.py (hunk 1 above)
Comment (felixfontein) [reply plugins/modules/redfish_command.py:901]: This is generally OK. Having return values with more or less complex content is normal for modules :)

I'm not sure the name `redfish_facts` fits. Maybe use something more general like `return_values` instead? `_facts` sounds a bit like `ansible_facts`, and these are not Ansible facts.

You should also document this return value in https://github.com/ansible-collections/community.general/blob/e152a1556fb1ade1d25ed339619d21b1951884ea/plugins/modules/redfish_command.py#L628-L634.

File: plugins/modules/redfish_command.py [hunk 3]
```diff
@@ -161,6 +161,18 @@
         description:
           - Password for retrieving the update image.
         type: str
+  update_apply_time:
+    required: false
+    description:
+      - Time when to apply the update.
+    type: str
```
[the remark this answers is not in the document]
Comment (mraineri) [reply plugins/modules/redfish_command.py:168]: Will expand the suggestion

File: plugins/modules/redfish_command.py (hunk 1 above)
Comment (mraineri) [reply plugins/modules/redfish_command.py]: Changed the name to `return_values` and added documentation about it.

File: plugins/modules/redfish_command.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mraineri) [reply plugins/modules/redfish_command.py:168]: Done

File: plugins/modules/redfish_command.py [hunk 4]
```diff
@@ -610,6 +637,19 @@
     returned: always
     type: str
     sample: "Action was successful"
+return_values:
+    description: Dictionary containing command-specific response data from the action
+    returned: on success
+    type: dict
```
Comment (felixfontein) [inline plugins/modules/redfish_command.py:643]:
```suggestion
    type: dict
    version_added: 6.1.0
```

File: plugins/modules/redfish_command.py [hunk 5]
```diff
@@ -610,6 +637,19 @@
     returned: always
     type: str
     sample: "Action was successful"
+return_values:
+    description: Dictionary containing command-specific response data from the action
```
Comment (felixfontein) [inline plugins/modules/redfish_command.py:641]:
```suggestion
    description: Dictionary containing command-specific response data from the action.
```

File: plugins/modules/redfish_info.py [hunk 6]
```diff
@@ -350,7 +366,8 @@ def main():
             username=dict(),
             password=dict(no_log=True),
             auth_token=dict(no_log=True),
-            timeout=dict(type='int', default=10)
+            timeout=dict(type='int', default=10),
+            update_handle=dict()
```
Comment (felixfontein) [inline plugins/modules/redfish_info.py:370]:
```suggestion
            update_handle=dict(),
```

**Review by felixfontein — APPROVED** [review]
Looks good from my side. When you're happy for this to be merged, tell me.

**Comment by mraineri** [comment]
> Looks good from my side. When you're happy for this to be merged, tell me.

Thanks! I'm happy with everything at this point. When I showed other folks the command name "PerformRequestedOperations", they suggested to leave it as is since it is accurate in what it's doing.

**Comment by felixfontein** [comment]
@mraineri thanks for your contribution!

**Pull request merged**